backend/src/models/ScenarioGraph.js

```diff
@@ -0,0 +1,123 @@
+/**
+ * Creates a graph representation of a list of scenes.
+ * @param scene - An array of scene objects
+ */
+export default class ScenarioGraph {
+
+    // Store graph as a modified adjacency list for easy indexing
+    graph = {};
+
+    constructor(scenes) {
+
+        // Iterate over each scene
+        scenes.forEach((scene) => {
+
+            let children = [];
+            let leaf = true;
+
+            // Iterate over each component of a scene
+            scene.components.forEach((component) => {
+                
+                // Check if the scene has any links to other scenes
+                if (component.type === 'BUTTON') {
+                    children.push(component.nextScene);
+                    leaf = false;    // Scene is not a leaf as it has a child.
+                }
+            })
+
+            // Add node to graph
+            this.graph[scene._id] = {
+                name: scene.name,
+                endScene: leaf,
+                linkedScenes: children
+            }
+        })
+    }
+
+    // Getters
+
+    get graph() { return this.graph }
+
+    // Methods
+
+    /**
+     * Computes the distance between two scenes.
+     * @param startScene: string - the id of the starting scene
+     * @param endScene: string - the id of the end scene
+     * @returns depth: number - the distance from the startScene to the endScene.
+     *                          If endScene does not exist within the graph or 
+     *                          there is not directed path from startScene to
+     *                          endScene, return -1.
+     */
+    distanceBetween(startScene, endScene) {

```

Comment:
Just a thought (no changes for now), but since the scene stays the same, we could actually generate a 2D distance map when the ScenarioGraph is constructed and save this data in like a 2D array (kinda like a 2D map).

Unsure about exact implementation, but constructor could build something similar to the `graph` field, but like...

`distances: Map<StartSceneId, EndSceneId => number>`

This means that provided this data model exists in the code (which in itself is related to caching in the backed – another story), we could get distances between any two scenes (which we will have to do every time the player moves to the next scene) without applying BFS n times.

Reply:
Yes! This is a much better solution I wish I had thought of. Definitely would improve runtime with O(1) indexing at the cost of O(V^2) space. Building the distance matrix would be O(V^3) with something like Floyd's algorithm, but would only happen when we initialize the graph object.

Looking back at my original implementation looks to be around O(V^3) (should have used something like Dijkstra), this would be a huge improvement. This would also help with server overhead if multiple people are playing scenarios.